Reapply #8644

aakselrod opened the pull request:
## Change Description
Fix #9229 by reapplying #8644 and
- correctly handling serialization errors in the `batch` package
- splitting `batch` requests into their own transactions for postgres db backend  to reduce serialization errors
- correctly handling errors that were previously ignored or not passed through in the `channeldb` package
- handling `current transaction is aborted` errors as serialization errors in case we hit a serialization error and ignore it, and get this error in a subsequent call to postgres
- tuning the `db-instance` postgres flags in `Makefile` per @djkazic's recommendations
- setting the `maxconnections` parameter for postgres DBs to 20 instead of 50 by default

## Steps to Test
See the failing itests prior to the fix, and the passing itests after the fix.

## Pull Request Checklist
### Testing
- [x] Your PR passes all CI checks.
- [x] Tests covering the positive and negative (error paths) are included.
- [x] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [x] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [x] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [x] Any new logging statements use an appropriate subsystem and logging level.
- [x] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [x] [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

aakselrod commented:
This push is going to fail. Pushing fix commit separately to demonstrate that it passes.

aakselrod commented:
Waiting to push fix commit until CI completes for the reapplication.

aakselrod commented:
Note the [postgres](https://github.com/lightningnetwork/lnd/actions/runs/11639430396/job/32415599077?pr=9242) [itests](https://github.com/lightningnetwork/lnd/actions/runs/11639430396/job/32415599244?pr=9242) fail mostly on opening, announcing, and closing channels. This is due to the use of the `batch` package to batch announcement update writes, while `batch` doesn't handle serialization errors correctly.

aakselrod commented:
Looks like there are still a couple of itests failing. Will keep working on this next week.

aakselrod commented:
The error message `not enough elements in RWConflictPool to record a read/write conflict` tells me that postgres might not be running with enough resources to handle the itests with 8644 reapplied.

Roasbeef commented:
This looks relevant, re some of the errors I see in the latest CI run: https://stackoverflow.com/a/42303225

Roasbeef commented:
Perhaps part of the issue is with the `ON CONFLICT` clause in many of the queries:
  * ON CONFLICT Clause with Partial Indexes: Your use of ON CONFLICT with a partial unique index and a WHERE clause may not be matching the index as expected. This can lead to unexpected behavior in conflict resolution and contribute to serialization failures.
  * Conflict Resolution: The ON CONFLICT clause relies on unique indexes or constraints to detect conflicts. If the clause doesn't perfectly match an existing unique index or constraint, PostgreSQL cannot efficiently perform conflict resolution, leading to increased chances of serialization failures.
  * Match ON CONFLICT Clause to Unique Index: Ensure that the ON CONFLICT clause exactly matches the unique index or constraint without additional WHERE clauses that might prevent PostgreSQL from recognizing the conflict.

Based on the SO link above, we might also be lacking some needed indexes.

aakselrod commented:
With closing the channel and a couple of other tests, I'm seeing logs similar to:

```
2024-11-01 22:13:15.503 [ERR] GRPH builder.go:1020: unable to prune routing table: unknown postgres error: ERROR: current transaction is aborted, commands ignored until end of transaction block (SQLSTATE 25P02)
2024-11-01 22:13:15.504 [ERR] GRPH builder.go:849: unable to prune graph with closed channels: unknown postgres error: ERROR: current transaction is aborted, commands ignored until end of transaction block (SQLSTATE 25P02)
```

when I reproduce locally. I'm going to pull on that thread first...

Roasbeef commented:
> 2024-11-01 22:13:15.504 [ERR] GRPH builder.go:849: unable to prune graph with closed channels

This looks like a case where we `continue` when we get an error, instead of checking it for a serialization error, and returning if it is.

aakselrod commented:
> This looks like a case where we `continue` when we get an error, instead of checking it for a serialization error, and returning if it is.

Yep, looking into why that isn't caught by the panic/recover mechanism.

aakselrod commented:
It was actually a lack of error checking in `delChannelEdgeUnsafe()` when calling `updateEdgePolicyDisabledIndex()`. This caused the serialization error to be ignored, and the next error is the one I pasted above. Submitting a fix commit below.

aakselrod commented:
Looks better as far as the errors on closing channels. Will keep working tomorrow to eliminate the other errors.

Roasbeef commented:
> 2024-11-05 03:38:04.408 UTC [14305] ERROR:  out of shared memory
> 2024-11-05 03:38:04.408 UTC [14305] HINT:  You might need to increase max_pred_locks_per_transaction.

Hmm, so we don't have great visibility into how much memory these CI machines have. Perhaps we need to modify the connection settings to reduce the number of active connections, and also tune params like `max_pred_locks_per_transaction`. 

@djkazic has been working on a postgres+lnd tuning/perf guide, that I think we can eventually check directly into lnd.

Roasbeef commented:
This is also very funky: https://github.com/lightningnetwork/lnd/blob/e3cc4d70c037d61e70b900ae69a9a8f388492c9c/kvdb/sqlbase/readwrite_bucket.go#L336-L363

We do two queries to just delete: select to see if exists, then delete. Instead of just trying to delete. 

Stepping back a minute: perhaps the issue is with this flawed KV abstraction we have. Perhaps we should just re-create a better hierarchical KV table from scratch. We use `sqlc` elsewhere so we can gain by having a unified set of light abstractions over what we want to do on the SQL layer.

Roasbeef commented:
Here's another instance of duplicated work in `CreateBucket`: https://github.com/lightningnetwork/lnd/blob/e3cc4d70c037d61e70b900ae69a9a8f388492c9c/kvdb/sqlbase/readwrite_bucket.go#L149-L187

We select to see if it exists, then potentially do the insert again. Instead, we can just do an `UPSERT`, then use `RETURNING` to give us the bucket id and key we care about, so a single query.

aakselrod commented:
I've been able to reduce (but not fully eliminate) the `out of shared memory` and `not enough elements in RWConflictPool` errors locally by changing the `-N` parameter of the `lnd-postgres` container to 200, and changing the default `maxconnections` value to 20. This follows from [this comment](https://github.com/postgres/postgres/blob/f638aafd1ea8505750ca9546e0e0b2e4cf31027d/src/backend/storage/lmgr/predicate.c#L1290-L1300) about how the `RWConflictPool` is allocated.

I've also tried treating these errors and `current transaction is aborted` as serialization errors, since they generally happen when too many transactions are conflicting, and that seemed to reduce the number of test failures.

In addition, I've found one more place where we get the `current transaction is aborted` errors due to lack of error handling, and added error handling there.

I'm pushing these changes for discussion below. My next step is to try to reduce the number of conflicts based on @Roasbeef's suggestions above. I'm going on vacation for the rest of the week until next Tuesday, so will keep working on this then.

aakselrod commented:
I think treating the OOM errors as serialization errors ended up being a mistake. Going to take that out and push when this run is done. In addition, I'm trying doubling the `max_pred_locks_per_transaction` value from the default ([64](https://www.postgresql.org/docs/13/runtime-config-locks.html)->128).

aakselrod commented:
Got some errors like:
```
ERROR:  could not access status of transaction
DETAIL:  Could not open file "pg_serial/0002": No such file or directory.
...
LOG:  could not truncate directory "pg_serial": apparent wraparound
```

Looking into those, they could also be from my change of treating the OOM errors as retryable.

aakselrod commented:
Set `-N 500` again while leaving the default per-node max connections at 20. This seems to improve things. Now starting work on reducing the number of actual serialization errors per suggestions above, rather than just optimizing settings/fixing our responses to them.

aakselrod commented:
Only 1 test failure last run, looks like no instances of `RWConflictPool` or `out of shared memory` but one instance of `pg_serial` right *after* the failure. Looking more closely into this.

djkazic commented:
> I think treating the OOM errors as serialization errors ended up being a mistake. Going to take that out and push when this run is done. In addition, I'm trying doubling the `max_pred_locks_per_transaction` value from the default ([64](https://www.postgresql.org/docs/13/runtime-config-locks.html)->128).

I'm running with

```
max_locks_per_transaction = 128
max_pred_locks_per_transaction = 1024
jit = off
checkpoint_timeout = 10min
work_mem = 8MB
```

Roasbeef commented on Makefile:
```diff
@@ -194,10 +194,10 @@ ifeq ($(dbbackend),postgres)
 	docker rm lnd-postgres --force || echo "Starting new postgres container"
 
-	# Start a fresh postgres instance. Allow a maximum of 200 connections so
+	# Start a fresh postgres instance. Allow a maximum of 500 connections so
```
Alternatively, we can also limit `sql` package params for lnd as well.

aakselrod replied:
Ah this was just me setting it back to the original config. In the end, I've been running tests locally with the existing config plus the parameters suggested by @djkazic above. I did set the default max connections to in LND. Will push cleaned up/latest code tomorrow.

aakselrod commented:
Looking at the last failure, I think it's a flake that's unrelated to the DB (we encountered it in our tests often before I added a hacky fix locally, it's due to an issue in txnotifier). When I run locally, I get some test framework related issues restarting nodes that I think I have a fix for, but I'll finish writing it and have the PR cleaned up tomorrow. Otherwise, I think it's generally passing tests despite the serialization failures.

aakselrod commented:
I'm still getting excessive serialization errors causing failures. Looks like that's hopefully the last place to fix. I'm trying out some schema/query changes, and it's likely to be pretty different than the existing schema. I'm going to try to do it all in the `kvdb/sqlbase` for the purpose of testing the changes for now. If/when I can get it to work better, we can talk about migration support.

aakselrod commented:
I have a somewhat incomplete attempt at rebuilding the KV SQL but haven't fixed the failures.

One possibly irrelevant thing I'm noticing is that both locally and in itests, with both my SQL refactor and before, the test failures all happen after testing 21 cases. Not sure where that's coming from or if it matters more than the number of serialization failures. Will keep looking...

aakselrod commented:
Pushed last update for the week:
- Updated postgres from v13 to v17
- Stopped batching multiple requests per tx in postgres in the `batch` package, and instead use 1 tx per request
- Removed the SQL schema/query changes for now as they don't seem to make much difference

I've had better results with those, but am bottlenecked by how long it takes the itests to run in terms of testing which changes are making the most difference.

aakselrod commented:
Looking into the failure above, I can reproduce it every few times running the test by itself locally with the postgres backend. It looks like we get 6 settle and 4 forward events rather than 5 forward and 5 settle. Working on figuring out why.

aakselrod commented:
Latest failure looks due to this error:

```2024-11-18 23:39:50.370 [ERR] GRPH builder.go:719: process network updates got: unable to add edge: unable to create shell node for: 035ffc6f909b451176d33b9b71d28859a26794a3bdf8e13006a6c0a3621633d803```

This seems to be due to the fact that the errors returned by the function adding the shell node don't pass through the underlying DB error, making it impossible to detect the serialization error and retry. Pushing a fix commit.

aakselrod commented:
Looking to see if the sqlite unit test failure was caused by this

aakselrod commented:
- The nativesql postgres test failure was cause by RWConflictPool. I've readded that to the list of serialization errors for retries, but only for text parsing so it doesn't include the "out of shared memory" errors. Pushing fix commit shortly.
- The kvdb postgres test failure for "reject onward htlc" was caused by `db tx retries exceeded`. I'm increasing the max retry parameter, but it may be worthwhile to make it configurable.
- The kvdb postgres test failure for "revoked uncooperative close retribution" was also caused by `db tx retries exceeded`
- The kvdb postgres test failure for "htlc timeout resolver extract preimage remote" was I think caused by an unrelated flake.

aakselrod commented:
Only failure this time was "htlc timeout resolver extract preimage" - I'll see if I can figure out why it happens tomorrow

aakselrod commented:
It seems to be reproducible locally for me, should be able to have it fixed today.

aakselrod commented:
So I think there's an issue here with the fact that `btcwallet` requests locks inside transactions, and then re-requests them on retry. When the inner function panics due to a serialization error, the lock is never released, so the retry just tries acquiring the lock when it's already held, causing a deadlock.

I think *for now* the answer is to keep a global lock facility and only turn it on for `walletdb` and then fix the issue in `btcwallet` separately. Working on this fix now.

Roasbeef commented:
> When the inner function panics due to a serialization error, the lock is never released, so the retry just tries acquiring the lock when it's already held

Why does this inner function panic? Is this another instance where we aren't properly catching the error? Or is it that we actually have panics in `kvdb` that trigger when errors fall through normal pattern matching?

Roasbeef commented:
Ah ok, I see now that the serialization error handling in general is based around recovering after panics to retry a transaction: https://github.com/lightningnetwork/lnd/blob/a10195042e7d3ceb138fdd529690466d5358b4d3/kvdb/sqlbase/db.go#L168-L198

https://github.com/lightningnetwork/lnd/blob/a10195042e7d3ceb138fdd529690466d5358b4d3/kvdb/sqlbase/db.go#L238-L246

I'm not sure why we went in that direction historically. At this kvdb emulation level, we can just pass through that error (not panic), then rely on the normal serialization error handling: https://github.com/lightningnetwork/lnd/blob/a10195042e7d3ceb138fdd529690466d5358b4d3/sqldb/interfaces.go#L268-L284

aakselrod commented:
I'll be off tomorrow but I'll see if I can refactor this to avoid panics. I think the biggest reason is that `Get()` doesn't return an error and is usually called within transactions.

Roasbeef commented:
Looks like we have a clean run `bitcoind-postgres` with the latest push? Nice work! 

Will check out the set of core commits now. I still think we can likely re structure the the queries and KV-table, but we can save that for another time.

Roasbeef commented on sqldb/sqlerrors.go:
```diff
@@ -17,6 +17,11 @@ var (
 	// than the max allowed valued without a success.
 	ErrRetriesExceeded = errors.New("db tx retries exceeded")
+
+	postgresErrMsgs = []string{
```
Style nit: missing `godoc` comment.

Roasbeef commented on sqldb/sqlerrors.go:
```diff
@@ -21,6 +21,7 @@ var (
 	postgresErrMsgs = []string{
 		"could not serialize access",
 		"current transaction is aborted",
+		"not enough elements in RWConflictPool",
```
IIUC, this gets returned when the instance [runs out of shared memory](https://www.postgresql.org/message-id/4E202D52020000250003F3A1@gw.wicourts.gov). I wager this is popping up mainly due to the constrained environment that the CI runners execute on.

Roasbeef commented on lncfg/db.go:
```diff
@@ -38,7 +38,7 @@ const (
 	DefaultBatchCommitInterval = 500 * time.Millisecond
 
-	defaultPostgresMaxConnections = 50
+	defaultPostgresMaxConnections = 20
```
With the other fixes later in this commit, if we revert this (back to 50) do things still pass? 

If not, then we may want to implement clamping for an upper limit here.

Roasbeef commented on Makefile:
```diff
@@ -200,7 +200,7 @@ ifeq ($(dbbackend),postgres)
 	# production nodes.
 	docker run --name lnd-postgres -e POSTGRES_PASSWORD=postgres -p 6432:5432 -d postgres:13-alpine -N 500 -c max_pred_locks_per_transaction=1024 -c max_locks_per_transaction=128 -c jit=off -c work_mem=8MB -c checkpoint_timeout=10min -c enable_seqscan=off
-	docker logs -f lnd-postgres &
+	docker logs -f lnd-postgres >itest/postgres-log 2>&1 &
```
Should this be reverted? 

Or looks like it just avoids printing out some of the retry spam.

Roasbeef commented on batch/batch_postgres.go:
```diff
@@ -0,0 +1,44 @@
+
+	// Apply each request in the batch in its own transaction. Requests that
+	// fail will be retried by the caller.
+	for _, req := range b.reqs {
```
Perhaps we should just bypass the batch scheduler all together for postgres? 

IIRC, we added it originally to speed up initial graph sync for bbolt, by reducing the number of total transactions we did.

Roasbeef commented on kvdb/sqlbase/db.go:
```diff
@@ -234,7 +202,19 @@ func (db *db) executeTransaction(f func(tx walletdb.ReadWriteTx) error,
-		return catchPanic(func() error { return f(kvTx) })
+
+		err := f(kvTx)
+		// Return the internal error first in case we need to retry and
```
Style nit: missing a newline above.

Roasbeef commented on sweep/sweeper.go:
```diff
@@ -1624,14 +1625,25 @@ func (s *UtxoSweeper) monitorFeeBumpResult(resultChan <-chan *BumpResult) {
 		case <-s.quit:
-			log.Debugf("Sweeper shutting down, exit fee " +
-				"bump handler")
+			log.Debugf("Sweeper shutting down, exit fee "+
```
Temp commit that can be dropped?

Roasbeef reviewed (commented):
Nice work with this PR! I can tell some serious tenacity went into iterating on this PR to get to the point it's at now.

aakselrod reviewed (commented):
I'll clean this up shortly, but responding to a few comments.

Note that the failure in CI from the previous push is the same deadlock we've seen before in `htlc_timeout_resolver_extract_preimage_(remote|local)` but in a different test, so it didn't end up showing the goroutine dump. But I think I've tracked it down and will submit a PR to btcwallet to fix it.

Roasbeef commented:
Re the shared memory issue, I think we can get around that by bumping up the size of the CI instance we use for these postgres tests: https://docs.github.com/en/actions/using-github-hosted-runners/using-larger-runners/running-jobs-on-larger-runners

aakselrod commented:
I think I've found the deadlock. With more than one DB transaction allowed in parallel for btcwallet, we're running into a deadlock similar to the following. This example is from the UTXO sweeper tests, but it can happen in other situations as well.

In one goroutine, the UTXO sweeper calls `NewAddress()` on the wallet. This:
- Starts a DB transaction
- Calls `(*wallet.wallet) newAddress()`, which
- Calls `FetchScopedKeyManager()` on the `waddrmgr.Manager`, which
- Calls `RLock()` and `RUnlock()` on the `waddrmgr.Manager` mutex and returns the requested scoped manager, holding no locks
- Then it calls `NextExternalAddresses()` on the `waddrmgr.ScopedManager`, which
- Calls `Lock()` on the `ScopedManager`'s mutex, defers the `Unlock(), and
- Calls `nextAddresses()` on the ScopedManager, which
- Calls `WatchOnly()` on the parent `Manager`, which
- Calls `RLock()` and `RUnlock()` on the `Manager`'s mutex

So while the top-level `Manager`'s mutex is locked first, it's immediately unlocked, and then *the `ScopedManager` lock is held and then `Manager` is locked/unlocked*

In another transaction running in parallel, we see the sweeper call `GetTransactionDetails()` on the wallet, which eventually:
- Starts a DB transaction
- Within that transaction, calls `(*waddrmgr.Manager) Address()`, which
- Calls `RLock()` on the `Manager`, and defers the `RUnlock()`
- Iterates through each `ScopedManager` to look for a matching address, which
- Calls `RLock()` and then `RUnlock()` on the `ScopedManager` to check if the requested address is cached, and if not
- Calls `Lock()` and then `Unlock()` on the `ScopedManager` to try to load/cache the requested address from the DB

So the sequence in this case is that *the `Manager` lock is held and the `ScopedManager`s are locked/unlocked*.

This has previously been mitigated by the fact that each of these happens inside a database transaction, which never ran in parallel. However, with parallel DB transactions made possible by this change, the inner deadlock is exposed.

I'll submit a PR next week to btcwallet to fix this, and then clean up this PR/respond to the comments above.

aakselrod commented:
I've submitted https://github.com/btcsuite/btcwallet/pull/967 to fix the deadlock mentioned above.

yyforyongyu commented:
> > Within that transaction, calls (*waddrmgr.Manager) Address(), which

Could you share the code line that does this? Tried to find it in `GetTransactionDetails` but no luck.

yyforyongyu commented on Makefile:
```diff
@@ -195,10 +195,12 @@ ifeq ($(dbbackend),postgres)
+	# each can run concurrently. Note that many of the settings here are
+	# specifically for integration testing and are not fit for running
+	# production nodes.
+	docker run --name lnd-postgres -e POSTGRES_PASSWORD=postgres -p 6432:5432 -d postgres:13-alpine -N 500 -c max_pred_locks_per_transaction=1024 -c max_locks_per_transaction=128 -c jit=off -c work_mem=8MB -c checkpoint_timeout=10min -c enable_seqscan=off
```
I just realized we are using a single postgres instance for all the tranches...this is probably wrong as we should use independent DB and chain backend for each tranche as they are supposed to be isolated...I need to think about how to fix this.

yyforyongyu replied:
yeah totally, no need to block on this since it's pre-existing.

yyforyongyu commented on Makefile:
```diff
@@ -195,10 +195,12 @@ ifeq ($(dbbackend),postgres)
+	# specifically for integration testing and are not fit for running
+	# production nodes.
+	docker run --name lnd-postgres -e POSTGRES_PASSWORD=postgres -p 6432:5432 -d postgres:13-alpine -N 500 -c max_pred_locks_per_transaction=1024 -c max_locks_per_transaction=128 -c jit=off -c work_mem=8MB -c checkpoint_timeout=10min -c enable_seqscan=off
+	docker logs -f lnd-postgres >itest/postgres-log 2>&1 &
```
can we name this `postgres.log` instead and add it to the gitignore

yyforyongyu replied:
cool yeah let's fix it when we move DB instance per tranche, just FYI we usually do logs cleanup here,
https://github.com/lightningnetwork/lnd/blob/c8cfa59316d827377ba80b7cd1295f678068baf1/lntest/node/harness_node.go#L669-L674

yyforyongyu reviewed (commented):
Love all the in-depth analysis. Just wanna mention it may be helpful to run this on top of #9260 to reduce some of the noise created from the itest flakes so the results can be more indicative. I ran it once [here](https://github.com/yyforyongyu/lnd/actions/runs/12025002336/job/33521422813?pr=18) and got a failure in postgres DB backend.

The other thing is I didn't realize how easily it is to trigger a serialization error in postgres. By simply running `list_channels` locally,
```
make itest icase=list_channels backend=bitcoind dbbackend=postgres nativesql=true
```
I got a concurrent update in `channeldb`,
```log
2024-11-26 11:13:39.317 UTC [89] STATEMENT:  INSERT INTO channeldb_kv (key, value, parent_id) VALUES($1, $2, $3) ON CONFLICT (key, parent_id) WHERE parent_id IS NOT NULL DO UPDATE SET value=$2 WHERE channeldb_kv.value IS NOT NULL
2024-11-26 11:13:39.379 UTC [62] ERROR:  could not serialize access due to concurrent update
```

aakselrod commented:
> > > Within that transaction, calls (*waddrmgr.Manager) Address(), which
> 
> Could you share the code line that does this? Tried to find it in `GetTransactionDetails` but no luck.

Sure:

`GetTransactionDetails` calls the underlying `GetTransaction`: https://github.com/lightningnetwork/lnd/blob/fbeab726e17bca3788c084ecb081a4699dd70091/lnwallet/btcwallet/btcwallet.go#L1375

`GetTransaction` starts a DB transaction: https://github.com/btcsuite/btcwallet/blob/66a3aeef6e78751e644a3d03fd856e982e0db4ac/wallet/wallet.go#L2543

Inside that, it calls `makeTxSummary`: https://github.com/btcsuite/btcwallet/blob/66a3aeef6e78751e644a3d03fd856e982e0db4ac/wallet/wallet.go#L2557

`makeTxSummary` calls `lookupOutputChain`: https://github.com/btcsuite/btcwallet/blob/66a3aeef6e78751e644a3d03fd856e982e0db4ac/wallet/notifications.go#L135

`lookupOutputChain` calls `(*waddrmgr.Manager) Address`: https://github.com/btcsuite/btcwallet/blob/66a3aeef6e78751e644a3d03fd856e982e0db4ac/wallet/notifications.go#L87

aakselrod commented:
I've cleaned this up a bit, responded to most of the comments in the code, and am now testing the DB tuning some more (maxconnections setting vs. container startup flags). Also testing eliminating one or two potentially extraneous commits to see if things still work after the latest bugfixes I've added.

Once I have that working, I'll rebase on #9260 and force-push, hopefully later this afternoon. I'd still expect failures on blockchain sync until btcsuite/btcwallet#967 is merged/tagged and I can reference it from `go.mod`, but otherwise that seems to fix the final deadlock we saw in CI here.

Looking forward to winding this up and getting it merged!

aakselrod commented:
Huh, all the itests failed. It seems to be working for me locally. Looking into it...

aakselrod commented:
I've rebased on `master` here. Opening a new PR and rebasing on #9260.

aakselrod commented:
Doing updates in #9313 now, though I can go back to here if that's preferred.

aakselrod commented:
Tests in #9313 are passing. Should I backport to this PR so the changes are based on `master`? There weren't many changes.

aakselrod commented:
Backported fixes from #9313 and rebased on latest `master`.

yyforyongyu commented on go.mod:
```diff
@@ -203,6 +203,9 @@ replace github.com/ulikunitz/xz => github.com/ulikunitz/xz v0.5.11
 replace github.com/gogo/protobuf => github.com/gogo/protobuf v1.3.2
 
+// Use local kvdb package until new version is tagged.
+replace github.com/lightningnetwork/lnd/kvdb => ./kvdb
```
Is this tagged now?

aakselrod replied:
No, the following commit in this PR modifies `kvdb/postgres` and will have to be tagged after it's merged. Then it will be safe to remove the redirect. Same goes for sqldb.

yyforyongyu commented on go.mod:
```diff
@@ -203,6 +203,9 @@ replace github.com/ulikunitz/xz => github.com/ulikunitz/xz v0.5.11
 // https://deps.dev/advisory/OSV/GO-2021-0053?from=%2Fgo%2Fgithub.com%252Fgogo%252Fprotobuf%2Fv1.3.1
 replace github.com/gogo/protobuf => github.com/gogo/protobuf v1.3.2
 
+// Use local kvdb package until new version is tagged.
```
Looks like this commit fab8386b2c0cd18e041993d64497905cddde4520 is missing?

aakselrod replied:
Oh I accidentally left off the commit to update btcwallet during rebase, I'll readd it.

aakselrod replied:
I've added a commit to update the `btcwallet` dependency again to eliminate the deadlock in `waddrmgr`. Thanks for noticing it!

yyforyongyu commented on Makefile:
```diff
@@ -195,9 +195,18 @@ ifeq ($(dbbackend),postgres)
+	# queries to lock individual rows instead of entire tables, helping
+	# reduce serialization conflicts. Disabling sequential scan for small
+	# tables also helps prevent serialization conflicts by ensuring lookups
+	# lock only relevant rows in the index rather than the entire table.
```
Nice doc!

yyforyongyu reviewed (commented):
Confirm this fixes the postgres failures on top of #9260! Have some questions about the tagging, otherwise think it's good to go.

aakselrod reviewed (commented):
Getting there!

aakselrod commented:
Eek, I have #9313 in the release notes instead of #9242. Going to fixup and force push.

aakselrod commented:
Guess I should rebase again.

yyforyongyu commented:
hmm I'm seeing this `commit unexpectedly resulted in rollback` from the logs
```
2024-12-10 19:57:55.524 [ERR] CRTR payment_lifecycle.go:197: Payment 51f23c9da1904012b48b2c70d3a5e6e9dd1f63189994beca5580d229b919ef5f with status=&{0 50000000 mSAT 0 mSAT false false} failed: commit unexpectedly resulted in rollback
2024-12-10 19:57:55.525 [ERR] CRTR router.go:989: Payment 51f23c9da1904012b48b2c70d3a5e6e9dd1f63189994beca5580d229b919ef5f failed: commit unexpectedly resulted in rollback
```

which is found from a [build](https://github.com/yyforyongyu/lnd/actions/runs/12263000284/job/34213597876?pr=24), where I rebased this branch on top of the itest fix PR.

aakselrod commented:
Thanks! I'll fix that today and rebase again.

aakselrod commented:
Rebased and added logging for `kvdb/sqlbase`.

yyforyongyu reviewed (approved):
LGTM👍 Thanks for all the investigations and fixes!

aakselrod commented on batch/batch_test.go:
```diff
@@ -0,0 +1,74 @@
+
+	var (
+		mu     sync.Mutex
+		called int
```
I used a mutex because the underlying `TimeScheduler` already provides a facility to do so, but I can pass a nil in its place and use an atomic if that's what people prefer? I don't have a strong opinion.

yyforyongyu replied:
it's a nit-picking, so nvm

Roasbeef commented on go.mod:
```diff
@@ -203,6 +203,9 @@ replace github.com/ulikunitz/xz => github.com/ulikunitz/xz v0.5.11
 replace github.com/gogo/protobuf => github.com/gogo/protobuf v1.3.2
 
+// Use local kvdb package until new version is tagged.
+replace github.com/lightningnetwork/lnd/kvdb => ./kvdb
```
Should we merge the changes to `kvdb`, then tag, then merge this PR?

Roasbeef commented on kvdb/go.mod:
```diff
@@ -135,6 +136,9 @@ replace github.com/dgrijalva/jwt-go => github.com/golang-jwt/jwt v3.2.1+incompat
 replace github.com/ulikunitz/xz => github.com/ulikunitz/xz v0.5.11
 
+// Use local sqldb package until new version is tagged.
+replace github.com/lightningnetwork/lnd/sqldb => ../sqldb
```
Leaving comment so we remember the replace is here.

Roasbeef reviewed (approved):
LGTM 🦆

Well done re your diligence and tenacity with this PR saga @aakselrod! I think we can leave some of those other SQL statement/schema level optimizations for another time. 

Only remaining question is what should we do about merge order here? We have 2 modules that were updated, and I think 3 replaces in this PR. Should we just merge as is, then make another PR to remove the replaces after a tag? Or do incremental merges and tags, with this as the final merge?

aakselrod commented:
@Roasbeef Thanks! I think the merge order is up to you and I'm happy to do it either way.

My input is that I *prefer* to merge the whole PR and then tag/remove the replaces, because I don't like merging code changes that are hidden until they're activated by tag. This way, the entire code change goes in together, its total functionality is obvious, and the tagging/replace removal after the fact won't change the underlying code at all. This also helps avoid making any mistakes in refactoring.

But if you prefer, I'm still happy to take out the replaces and refactor to 2-3 PRs, where the code changes for the first 1-2 are hidden until tagged. I see the advantage of not messing around with the `go.mod` files and replaces so much.

aakselrod commented:
Made one quick change to release notes (added my name) in case you decide to merge this PR as-is. Sorry!

yyforyongyu commented:
Think we can merge this and tag, then I will clean up #9367 to remove the replaces.
